core/src/main/java/com/linecorp/armeria/internal/common/AbstractHttp2ConnectionHandler.java

```diff
@@ -162,8 +170,16 @@ public final ChannelFuture goAway(ChannelHandlerContext ctx, int lastStreamId, l
         return super.goAway(ctx, lastStreamId, errorCode, debugData, promise);
     }
 
+    protected final void setGoAwayDebugMessage(String debugMessage) {
+        goAwayDebugData = Unpooled.copiedBuffer(debugMessage.getBytes()).asReadOnly();
+    }
+
+    protected final ChannelFuture goAway(ChannelHandlerContext ctx, int lastStreamId) {
+        return goAway(ctx, lastStreamId, NO_ERROR.code(), goAwayDebugData.retain(), ctx.newPromise());
```

Comment:
Once we make `goAwayDebugData` unreleasable above, `.retain()` is unnecessary here.